### ISSUE TITLE
Memory leak activated when setting a seed in TensorFlow 2 (2.0 & 2.1) and using eager execution.

### ISSUE THREAD

**Issue by IanRDavies** [tensorflow/tensorflow#37252]
Firstly, I would like to thank @condnsdmatters for his help in investigating this issue and tracing the problem back to the use of a random state when setting a seed for TensorFlow.

**System information** 
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.1.0 (also tested on 2.0.0)
- Python version: Python 3.7.5
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: - GPU model and memory: N/A

**Describe the current behaviour**
When a random seed is set in TensorFlow 2 (2.0 or 2.1) memory usage increases with every call to usage of any random sampling in TensorFlow. We believe that this is due to the use of a Python's `random.Random` random state as a source of randomness and a lack of garbage collection for the results of random sampling. We may see that when a seed is set TensorFlow uses `random.Random` by looking at the [`_set_global_seed`](https://github.com/tensorflow/tensorflow/blob/e5bf8de410005de06a7ff5393fafdf832ef1d4ad/tensorflow/python/eager/context.py#L433) method of an eager execution context.

The reproduction example does not use `tf.random.set_seed` since we have tracked the issue down more narrowly to the sampling of random values and we therefore concentrate on a minimal focussed example in the hope of enabling a faster fix.

See the reproduction below and [our colab notebook](https://colab.research.google.com/drive/1HhVduWlY6Va_KLX2b-SqCGIWLNDy5MFT) for concrete examples of when this occurs. See the note at the end of this post before running the code.

**Describe the expected behaviour**
We would not expect memory usage to grow with the number of calls to random sampling. In the example below we would expect constant memory usage.

**Standalone code to reproduce the issue** 
The code below will show increasing memory usage over time.

```python
import tensorflow as tf
from random import Random
import os
import psutil

MAXINT = 2**32 - 1

def print_memory_usage():
    print("memory used:", psutil.Process(os.getpid()).memory_info().rss)
    return psutil.Process(os.getpid()).memory_info().rss

if __name__ == "__main__":
    random_state = Random()
    for _ in range(10):
        for _ in range(1000):
            a = random_state.randint(0, MAXINT)
            tf.Variable(a)
        print_memory_usage()
```

Further examples of cases with and without memory leaks are provided in [this colab notebook](https://colab.research.google.com/drive/1HhVduWlY6Va_KLX2b-SqCGIWLNDy5MFT).

**Other info / logs**
If the random integer sampled in the example is cast to a float the memory issue remains. However, if it is cast to a string before being passed to `tf.Variable` then there is no memory issue. This leads us to think that the issue is related to the memory usage as laid out in the C++ code of TensorFlow [here](https://github.com/tensorflow/tensorflow/blob/b87f4a54652d72435a088e5a510ca46e124deccd/tensorflow/c/tf_tensor.cc#L109).

This is an issue with eager execution as defining the loops in a function and decorating it with `tf.function` resolves the memory issue. This then leads us to the more targetted example included below.

```python
from random import Random
from tensorflow.python.framework.constant_op import convert_to_eager_tensor
from tensorflow.python.eager import context
import os
import psutil

MAXINT = 2**32 - 1

def print_memory_usage():
    print("memory used:", psutil.Process(os.getpid()).memory_info().rss)
    return psutil.Process(os.getpid()).memory_info().rss


if __name__ == "__main__":
    random_state = Random()
    ctx = context.context()
    for _ in range(10):
        for _ in range(1000):
            a = random_state.randint(0, MAXINT)
            convert_to_eager_tensor(a, ctx)

        print_memory_usage()
```

**Note:** Running the code in [the colab notebook](https://colab.research.google.com/drive/1HhVduWlY6Va_KLX2b-SqCGIWLNDy5MFT) provided can lead to inconsistent results due to the cloud execution environment. We suggest running our examples locally when debugging. At the very least the run time should be reset and each example run from a clean starting point before conclusions are drawn.

**Comment by foxik** [comment]
I believe this is caused by caching which `convert_to_eager_tensor` performs. Notably look at https://github.com/tensorflow/tensorflow/blob/7072568ed6b735e347fb87bc84a2b83daf806e3f/tensorflow/python/framework/constant_op.py#L68-L72 which mentions that some values can be cached.

In the C++ code, https://github.com/tensorflow/tensorflow/blob/7072568ed6b735e347fb87bc84a2b83daf806e3f/tensorflow/python/eager/pywrap_tensor.cc#L307-L313 performs the caching -- currently for Python scalars only.

This is why you need to generate random numbers -- each call allocates a new Eager tensor with that specific constant. If you instead use a constant, it allocates the eager tensor only once.

**Comment by IanRDavies** [comment]
Update: I reran the reproduction code with TensorFlow 2.2 and the issue is still present.

**Comment by gowthamkpr** [comment]
@IanRDavies Thanks for updating us

**Comment by IanRDavies** [comment]
I just had a quick review with the latest versions of TensorFlow. The behaviour in TensorFlow versions 2.3.2 and 2.4.1 is consistent with observations in past versions. The issue still exists.

**Comment by tilakrayal** [comment]
@IanRDavies,
Could you please check the attached [gist](https://colab.research.google.com/gist/tilakrayal/5b776e3a60e59b513ff5f9588d3c6238/memory_leak_in_tensorflow_2.ipynb) of the latest version and let us know if the issue still persists. Thank you!

**Comment by SnoopJ** [comment]
> Could you please check the attached [gist](https://colab.research.google.com/gist/tilakrayal/5b776e3a60e59b513ff5f9588d3c6238/memory_leak_in_tensorflow_2.ipynb) of the latest version and let us know if the issue still persists. Thank you!

I can confirm that this issue affects TensorFlow 2.9.0 (on Ubuntu 20.04, Python 3.9.14). The original reporter's program rapidly uses up all available GPU memory when I run it.

It is a terrible shame to see such a wonderfully thorough bug report auto-closed. It reflects very poorly on how TensorFlow views their community that helpful reports like this are discarded because they fail to catch a maintainer's attention, are marked "stale" because of that lack of attention, and are then thrown away.

**Comment by SnoopJ** [comment]
I can also confirm that this issue also manifests a steady growth in memory utilization on `tf-nightly==2.16.0.dev20231115` (Ubuntu 20.04, Python 3.12.0)

This problem is very much still impacting TensorFlow, please re-open this issue.

**Comment by mihaimaruseac** [comment]
Please open a new issue (optionally referencing this one), as this one has been for an old version of TF and there has been some work to alleviate this, so if this happens now too it's likely there was some regression or there's more work needed
